File: problems-adapt/0198_maximum-non-adjacent-loot/solution_dp.cpp
class Solution {
  public:
    int maxNonAdjacentLoot(vector<int> &nums) {
        // Rolling two-variable DP: cur = best through position i-1, prev = best
        // through position i-2; both start at 0 ("nothing taken yet").
        long long prev = 0, cur = 0;
        for (int x : nums) {
            // Skip this position (keep cur) or take it (prev + x).
            long long next = max(cur, prev + x);
            prev = cur;
            cur = next;
        }
        return (int)cur;
    }
};

File: problems-adapt/0001-0100/0068_maximum-non-adjacent-loot/solution_dp.go
func maxNonAdjacentLoot(nums []int) int {
	// Rolling two-variable DP: cur = best through position i-1, prev = best
	// through position i-2; both start at 0 ("nothing taken yet").
	prev, cur := 0, 0
	for _, x := range nums {
		// Skip this position (keep cur) or take it (prev + x); the tuple
		// assignment advances both values with no temporary array.
		prev, cur = cur, max(cur, prev+x)
	}
	return cur
}
